subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -0,0 +1,241 @@
+
+		return (ctx.cur_item_size <= ctx.buffer_size) ? 0 : -ENOMEM;
+	}
+
+	/* Parse CBOR header */
+	bool res;
+	uint_fast32_t image_count;
+
+	ZCBOR_STATE_D(states, CBOR_HEADER_NESTING_LEVEL, ctx.buffer + FIXED_HEADER_SIZE,
+		      ctx.cur_item_size - FIXED_HEADER_SIZE, 1);
+
+	res = zcbor_map_start_decode(states);
+	res = res && zcbor_tstr_expect_lit(states, "img");
+	res = res && zcbor_list_start_decode(states);
+	res = res && zcbor_multi_decode(1, CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT, &image_count,
+					(zcbor_decoder_t *)parse_image_info, states,
+					ctx.header.images, sizeof(struct image_info));
+	res = res && zcbor_list_end_decode(states);
+	res = res && zcbor_map_end_decode(states);
+
+	if (!res) {
+		return zcbor_pop_error(states);
+	}
+
+	ctx.header.image_count = image_count;
+	select_next_image();
+
+	return 0;
+}
+
+static const struct dfu_image_writer *current_image_writer(void)
+{
+	if (ctx.cur_image_no >= 0 && (size_t)ctx.cur_image_no < ctx.header.image_count) {
+		const int image_id = ctx.header.images[ctx.cur_image_no].id;
+
+		for (size_t i = 0; i < ctx.writer_count; i++) {
+			if (ctx.writers[i].image_id == image_id) {
+				return &ctx.writers[i];
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+{
+	int err = 0;
+
+	/* Process only remaining bytes of the current item (header or image) */
+	chunk_size = MIN(chunk_size, ctx.cur_item_size - ctx.cur_item_offset);
+
+	if (ctx.cur_image_no < 0) {
+		/* Parsing header */
+		memcpy(ctx.buffer + ctx.cur_item_offset, chunk, chunk_size);
+		ctx.cur_item_offset += chunk_size;
+
+		if (ctx.cur_item_offset == ctx.cur_item_size) {
+			err = parse_header();
+		}
+	} else {
+		/* Image data */
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (!writer) {
+			err = -ESPIPE;
+		}
+
+		if (!err && ctx.cur_item_offset == 0) {
+			err = writer->open(writer->image_id);
+		}
+
+		if (!err) {
+			err = writer->write(chunk, chunk_size);
+			ctx.cur_item_offset += chunk_size;
+		}
+
+		if (!err && ctx.cur_item_offset == ctx.cur_item_size) {
+			err = writer->close(true);
+			select_next_image();
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	return chunk_size;
+}
+
+int dfu_multi_image_init(uint8_t *buffer, size_t buffer_size, dfu_multi_image_seek_t seek)
+{
+	if (buffer == NULL || buffer_size < FIXED_HEADER_SIZE) {
+		return -EINVAL;
+	}
+
+	memset(&ctx, 0, sizeof(ctx));
+	ctx.buffer = buffer;
+	ctx.buffer_size = buffer_size;
+	ctx.seek = seek;
+	ctx.cur_image_no = -1;
+	ctx.cur_item_size = FIXED_HEADER_SIZE;
+
+	return 0;
+}
+
+int dfu_multi_image_register_writer(const struct dfu_image_writer *writer)
+{
+	if (ctx.writer_count == CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT) {
+		return -ENOMEM;
+	}
+
+	ctx.writers[ctx.writer_count++] = *writer;
+	return 0;
+}
+
+int dfu_multi_image_write(size_t offset, const uint8_t *chunk, size_t chunk_size)
+{
+	if (offset > ctx.cur_offset) {
```

Comment:
If the seek callback will be used, that will mean that `ctx.cur_offset` will have been adjusted accordingly. So it's a check against an unexpected data gap.

About the ordering... How would I then manage the state of the parser if chunks could come in an arbitrary order? What if I got a chunk for the next image before I closed the previous image? Should I drop it or buffer somewhere? But then what if I don't have space in the buffer? And actually why should we assume that first chunks containing the header come in the right order? I'm worried that if we start thinking about those cases, the library will get pretty complex and take a lot of RAM and flash to work properly. And many users of the library won't even use it, because they solve the ordering problem at a lower network layer (where it's supposed to be solved). Also, the DFU target library which we envisioned to be our default backend is not compatible with that approach, so I would need to actually implement reordering inside the library and I'm not sure it's a good idea :)

Let me know if you have an idea/algorithm in mind how you think it should work.

Reply:
I was talking about my feeling of this function from a user perspective. Out of order writes were something that felt like a valid scenario for this module, and I think that if I got this feeling, some customer who will want to modify this part may come to the same conclusion. When I read the documentation to this function, it seemed clear that it expects chunks in order, but not every one reads the documentation. (That is not an explanation to my oversight, just a suggestion how I perceived this interface in isolation to documentation) 


After some thought, I feel convinced that we should not implement this out of order writing.
In my opinion, to implement it we would have to enforce this capability on every writer, and we know that our writer will not have this feature. Furthermore, as You mentioned this should not be an issue, because those packets should be already arranged by other layers in the stack.

There is still issue of interface, that will lead programmers that they could do this.
I think if we started with file API, we should stick with it. As far as I remember, when I use seek on a file, it moves the read head of a file, and next `read(8)` will read the next 8 bytes of data from the point where head has been placed by seek. 

We could do the same here with write. I would propose to remove the offset from write, and if we get a seek callback, we will be forced to send in next write a chunk of data corresponding to that seek request. 

This approach would be more consistent with file API that we adopt, and it would not suggest the possibility of out of order writes.
We could document in the seek function that it will never go back, just to tell the users that they do not need to store any data that was missed by seek. 

In case of the protocols that can request specific chunk, this will result in asking the data provider for the data requested in seek callback, and in protocols where the image is just streamed, irrelevant data can be just ignored

This approach has its own flaws, like ignoring seek callback will result in complete chaos (writing next block that should be ignored in place of an expected data). Another issue is synchronization, If I get a stream of data, and I frequently call write with those data, and I get a seek callback just when I'm about to call next write. I could end up with incorrect write (I didn't check seek yet, and I already called write) 

Maybe we could solve this by using something like:
```
// signal that application received data
void data_ready();

// ask module what data it expects in next write
size_t payload_offset();

// callback to implement in application that will coppy data with requested offset to out buffor
// return number of bytes that were coppied
typeof size_t (*read_data)(size_t offset, char* out, size_t out_max_size);
```
In this scheme, the application may ask what chunk of data will be expected, fetch this from the data provider, and call data_ready. Next a read_data callback will be called, and offset parameter will ensure that the application copies only correct data (application need to keep track of the offset). If we add to this that read_data will be called in order, so that in subsequent calls offset will never decrease, it seems like a nice interface that allow arbitrary jumps forward.

P.S. I think that current implementation is acceptable, and I will not try to enforce my propositions. I just try to raise my concerns, and maybe in the process work out some better approach, and learn a few things ;)

Reply:
No worries about raising concerns. I just struggle to design an API that would meet all the requirements that come from different teams and I try to prioritize ease of use (after reading the module documentation :)).

Overall I get your point, but the `offset` argument was proposed during our last meeting to address two requirements:
- the API should allow the DFU library to skip some image on its own 
- reduce the boilerplate code.

If we removed the argument, and each transport that downloads fixed-sized blocks or downloads multiple blocks in parallel will need to crop the current chunk by the current offset before passing it to `dfu_multi_image_write()`. I simply wanted to avoid the burden for library users, but if we all agree (@tomchy?) I can remove the argument and shift the responsibility of cropping chunks onto the caller.

Reply:
Reworked the API according to our recent discussion. Please check again.